### PR TITLE
fix: avoid native tag as component

### LINKED ISSUE #123: Unexpected token? I haven't seen this before
4:01:13 PM [vite] Internal server error: Unexpected token (1:7)
  Plugin: unplugin-vue2-script-setup
  File: /Users/jimhurd/yakdb/aetnaticket/src/shell/BrowseTicket.vue
  1  |  <script lang="ts" setup>
     |         ^
  2  |  
  3  |  import { onMounted, reactive, ref, toRef, getCurrentInstance } from '@vue/composition-api'

If I reverse the template and the script it still finding a problem at an area that's clearly not a problem

4:08:16 PM [vite] Internal server error: Unexpected token (1:7)
  Plugin: unplugin-vue2-script-setup
  File: /Users/jimhurd/yakdb/aetnaticket/src/shell/BrowseTicket.vue
  1  |  <template>
     |         ^
  2  |    <div>
  3  |      <div class="chatEditPane" ref="chatPane" v-if="ticketId">

It's just this one file, If I stub out everything its fine.  Any tips or help you can offer is greatly appreciated. 

I did npm update I have
    "unplugin-vue2-script-setup": "^0.10.0",
    "vite": "^2.8.6",
    "vite-plugin-components": "^0.8.3",
    "vite-plugin-icons": "^0.4.0",
    "vite-plugin-vue2": "^1.4.2",
    "vue-template-compiler": "^2.6.14",
    "vue-tsc": "^0.31.1"

[BrowseTicket.vue.txt](https://github.com/antfu/unplugin-vue2-script-setup/files/8191501/BrowseTicket.vue.txt)

### PR DESCRIPTION
**Pull request by jaw52**
### Description

```vue
<script setup lang="ts">
let p = "not ok";
</script>

<template>
  <p>{{ p }}</p>
</template>
```
variable `p` will be misinterpreted as a component （Component names in vue avoid using html tag）

### Linked Issues

This closes https://github.com/antfu/unplugin-vue2-script-setup/issues/133 ,closes https://github.com/antfu/unplugin-vue2-script-setup/issues/135

### CODE REVIEW

**Comment by xiaoxiangmoe** [comment]
How about 
```vue
<script setup lang="ts">
import button from './button.vue'
</script>

<template>
  <button>{{ button }}</button>
</template>

**Comment by jaw52** [comment]
> 

This is the effect of the official scaffolding of` vite + vue3`

image

And. This is the effect of running after this commit
image

**Comment by jaw52** [comment]
> How about
> 
> ```
> <script setup lang="ts">
> import button from './button.vue'
> </script>
> 
> <template>
>   <button>{{ button }}</button>
> </template>
> ```

This is the effect of the official scaffolding of` vite + vue3`

image

And. This is the effect of running after this commit
image

**Comment by jaw52** [comment]
Do we need to add more test cases？

File: src/core/transformScriptSetup.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { capitalize } from '@vue/shared'
 import type { Node, ObjectExpression, Statement } from '@babel/types'
 import { notNullish, partition, uniq } from '@antfu/utils'
+import htmlTags from 'html-tags'
```
Comment (xiaoxiangmoe) [inline src/core/transformScriptSetup.ts:4]: try using `"@vue/compiler-dom`'s `parserOptions.isNativeTag`

Comment (jaw52) [reply src/core/transformScriptSetup.ts]: > try using `"@vue/compiler-dom`'s `parserOptions.isNativeTag`

That's a good idea. I've finished the unit tests~

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -2400,7 +2400,7 @@ packages:
       '@babel/types': 7.22.4
       '@vue/babel-helper-vue-transform-on': 1.0.2
       camelcase: 6.2.0
-      html-tags: 3.1.0
```
Comment (xiaoxiangmoe) [inline pnpm-lock.yaml:2403]: lockfile need't change

Comment (jaw52) [reply pnpm-lock.yaml]: > lockfile need't change

ok. Already restored

**Comment by jaw52** [comment]
Is there any work to be done about this PR @xiaoxiangmoe

File: test/__snapshots__/transform.test.ts.snap [hunk 3]
```diff
@@ -330,6 +330,72 @@ export default __sfc_main;
...
+`;
+
+exports[`transform > fixtures > test/fixtures/HtmlTag2.vue 1`] = `
+"<script lang=\\"ts\\">
+import Button from './DynamicStyle.vue';
+import button from './DynamicStyle.vue';
+import { defineAsyncComponent } from '@vue/composition-api';
+const footer = defineAsyncComponent(() => import('./ScriptOnly.vue'));
+import { ref } from '@vue/composition-api';
+const __sfc_main = {};
```
Comment (xiaoxiangmoe) [inline test/__snapshots__/transform.test.ts.snap:376]: This seems wrong in vue 3.
We should add `__sfc_main.components = {Button}`

see example:
https://play.vuejs.org/#eNqNUcFqwzAM/RXhSxMY9b1zAt3YroWxoy9d4mwBxza20jFC/n1y7aYttNBTLL0n6b2XiW2dWx9GxTZMhMb3DiEoHB3ovfmuJMMgWS1NPzjrEV5GRGug83aA1ZqnMo6vnhfO1wOcCVrV9UZtw59pXi01jTIIc56KZGkaawJCZy0qD9XNiaIooaohbS2ur5XlxTmvuqvtJCWtd1ARVkj2o7S28Gt9K1kJhAue8iD3VKAanN6jogpAtP3h+KCnq6eJtsyz4O7USxFEIAcW0dzMlARESs4rUnIzU5Lz+n23+3z7EDyXx/s8CRB8kcWe2Nn77Z+JIVk5+7rrKstect6ACNpiTbPxc0/D/A8NBMz/

File: src/core/transformScriptSetup.ts [hunk 4]
```diff
@@ -55,6 +56,7 @@ export function transformScriptSetup(
     })
 
   const components = Array.from(template.components)
+    .filter(component => !parserOptions.isNativeTag!(component.toLowerCase()))
```
Comment (xiaoxiangmoe) [inline src/core/transformScriptSetup.ts:59]: Why should it be `component.toLowerCase()`?

Comment (jaw52) [reply src/core/transformScriptSetup.ts]: Thank you brother, I have exported the original unprocessed `tags `for judgment

File: test/__snapshots__/transform.test.ts.snap (hunk 3 above)
Comment (jaw52) [reply test/__snapshots__/transform.test.ts.snap:412]: ok finished
Snipaste_2023-06-19_14-06-42

**Comment by jaw52** [comment]
@xiaoxiangmoe I'm ready to review, bro ^_^

File: playground/src/Foo.vue [hunk 5]
```diff
@@ -1,3 +1,8 @@
-<template>
-  <div>Foo</div>
-</template>
+<script setup lang=ts>
```
Comment (xiaoxiangmoe) [inline playground/src/Foo.vue:1]: Bug: `lang=ts`

File: playground/src/App.vue [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <script setup lang="ts">
 import { defineAsyncComponent } from '@vue/composition-api'
-
+import  ButtonTest  from './ButtonTest.vue';
```
Comment (xiaoxiangmoe) [inline playground/src/App.vue:3]: remove useless space here

File: test/fixtures/HtmlTag2.vue [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<script setup lang="ts">
+import Button from './DynamicStyle.vue';
+import button from './DynamicStyle.vue';
+import { defineAsyncComponent } from '@vue/composition-api'
+const footer = defineAsyncComponent(() => import('./ScriptOnly.vue'))
+
+import { ref } from '@vue/composition-api';
+
+const p =ref("hello word") ;
```
Comment (xiaoxiangmoe) [inline test/fixtures/HtmlTag2.vue:9]: add single space after `=`

File: src/types.ts [hunk 8]
```diff
@@ -1,63 +1,64 @@
...
+  found: boolean
+  ast: Program
+}
+
+export interface ParsedSFC {
+  id?: string
+  template: {
+    /** foo-bar -> FooBar */
+    components: Set<string>
+    tags: Set<string>
```
[the remark this answers is not in the document]
Comment (jaw52) [reply src/types.ts:21]: If 'template' contains' button ', 'components' will convert' button 'to' Button '.
I need to judge for `tags` that have not been converted

for example
```vue
<template>
    <button>TEST</button>
</template>
```
```js
console.log(template.components) //['Button ']
```

`template.components` returns `Button `.Using `parserOptions.isNativeTag` make false positives
```js
parserOptions.isNativeTag!('Button')//false
```

File: src/core/transformScriptSetup.ts [hunk 9]
```diff
@@ -1,228 +1,231 @@
...
+
+  // filter out identifiers that are used in `<template>`
+  const returns: ObjectExpression['properties'] = declarationArray
+    .filter(i => template.identifiers.has(i))
+    .map((i) => {
+      const id = t.identifier(i)
+      return t.objectProperty(id, id, false, true)
+    })
+
+  const components = Array.from(template.tags)
```
Comment (xiaoxiangmoe) [inline src/core/transformScriptSetup.ts:58]: What would here happen if `template.tags` is `['hello-world','HelloWorld']`

File: src/types.ts [hunk 10]
```diff
@@ -1,63 +1,64 @@
...
+  attrs: Record<string, string>
+  found: boolean
+  ast: Program
+}
+
+export interface ParsedSFC {
+  id?: string
+  template: {
+    /** foo-bar -> FooBar */
+    components: Set<string>
```
Comment (xiaoxiangmoe) [inline src/types.ts:20]: You can remove `components: Set<string>` here.

File: src/core/transformScriptSetup.ts
Comment (xiaoxiangmoe) [inline src/core/transformScriptSetup.ts:1]: image
This seems a little hard to review.
Can you squash all your commit? Thank you.

Comment (jaw52) [reply src/core/transformScriptSetup.ts:1]: sorry.My git parameters are wrong.Now corrected

File: src/core/transformScriptSetup.ts (hunk 9 above)
Comment (jaw52) [reply src/core/transformScriptSetup.ts]: > What would here happen if `template.tags` is `['hello-world','HelloWorld']`
I don't filter these non-native tags,The result should be the same as before

```js
//template.tags is ['hello-world','HelloWorld']
 const nonNativeTags = new Set(
    Array.from(template.tags)
      .filter(tag => !parserOptions.isNativeTag!(tag))
      .map(pascalize),
  )
  
console.log(nonNativeTags)//['hello-world','HelloWorld']

 const components = Array.from(nonNativeTags)
    .map(
      component =>
        declarationArray.find(declare => declare === component)
        ?? declarationArray.find(declare => pascalize(declare) === component),
    )
    .filter(notNullish)
```

File: test/fixtures/HtmlTag.vue
Comment (xiaoxiangmoe) [inline test/fixtures/HtmlTag.vue:1]: Please check weather all your file is using lf

Comment (jaw52) [reply test/fixtures/HtmlTag.vue:1]: I checked both locally and on github, and there seems to be no CRLF files

**Pull request merged**